OpenSearch Extensibility Roadmap

dblock opened the pull request:
### Description

Draft for #1408.
 
By submitting this pull request, I confirm that my contribution is made under the terms of the BSD-3-Clause License.

seanneumann reviewed (commented):
Is the filename supposed to be set to April 21st? Feels like it should be set to 03-21.

seanneumann commented on _posts/2023-04-21-extensibility-roadmap.markdown:
```diff
@@ -0,0 +1,154 @@
+
+We’ll augment the previously built rudimentary schema for extension’s metadata to provide additional fields beyond name, version or compatibility. These will include such fields as well-defined categories, or additional vendor/sponsor information. We will build a minimal catalog website with search functionality, deploy and maintain a public version of it. We’ll ensure that the catalog system can ...
+
+Developers will be able to publish an extension into a public catalog, users can find extensions in a catalog via a taxonomy and search functions, detail pages with meaningful metadata and vendor information, administrators can install extensions from public or private catalogs, and import a subset of a public catalog in their Enterprise. Finally, everyone we'll establish trust into who each exte ...
```
This sentence doesn't make sense to me....

"Finally, everyone we'll establish trust into who each extension publisher is claiming to be, and some quality controls and publisher approvals."

seanneumann commented on _posts/2023-04-21-extensibility-roadmap.markdown:
```diff
@@ -0,0 +1,154 @@
+
+Assuming extensions have been widely adopted, we can deprecate the plugin APIs to be removed in the next major version of OpenSearch. We don't expect this to happen earlier than OpenSearch 4.0. Older versions of plugins continue to work with older versions of OpenSearch and continue receiving security patches.
+
+We will replace the two distribution flavors of OpenSearch/Dashboards (currently a -min distribution without security or plugins) by a set of distribution manifests tailored for purposes such as log analytics vs. search. Each distribution will be represented by a manifest that can be assembled “on-the-fly” by downloading the artifacts from the public catalog for packaging purposes. We want to ena ...
```
If your intention is to key off of "enable", consider these changes...

"We want to enable single-click **installation of a collection** of extensions, recommended distributions of OpenSearch for tailored purposes, **the** ability for vendors to create their favorite flavor of OpenSearch/Dashboards distribution easily, **and the capability** to create enterprise-tailored distributions."

owaiskazi19 commented on _posts/2023-04-21-extensibility-roadmap.markdown:
```diff
@@ -0,0 +1,154 @@
+
+## Future
+
+Extensions make a lot of new, big ideas possible! Below are some of our favorites.
```
```suggestion
Extensions will make a lot of new, big ideas possible! Below are some of our favorites.
```

dblock replied:
Taken!

owaiskazi19 commented on _posts/2023-04-21-extensibility-roadmap.markdown:
```diff
@@ -0,0 +1,154 @@
+
+### Hot swap extensions 
+
+OpenSearch/Dashboards bootstrap plugins at start time and various parts of the system assume that plugins do not change at runtime. Requiring a cluster restart for all extensions is a severely crippling limitation on the path of ecosystem adoption of any significant number of extensions, primarily because cluster restarts mean stopping inbound traffic. We ensure that any part of the system can be ...
```
```suggestion
OpenSearch/Dashboards bootstrap plugins at start time and various parts of the system assume that plugins do not change at runtime. Requiring a cluster restart for all extensions is a severely crippling limitation on the path of ecosystem adoption of any significant number of extensions, primarily because cluster restarts mean stopping inbound traffic. We ensure that any extension point can be loaded or unloaded at runtime, starting with making settings in OpenSearch core dynamic, and add tooling to support loading and unloading extensions at runtime without restarting OpenSearch/Dashboards nodes or the entire cluster. This creates the ability to add, upgrade, or remove an extension without the need to restart OpenSearch/Dashboards, or connect a remote extension to an existing OpenSearch/Dashboards cluster.
```

dblock replied:
Taken!

owaiskazi19 commented on _posts/2023-04-21-extensibility-roadmap.markdown:
```diff
@@ -0,0 +1,154 @@
+
+### Extensions in other programming languages
+
+With extensions designed to operate remotely we can support out of process extensions written in other languages running on the JVM, and remote extensions written in any other language hosted externally. By enabling polyglot applications, or porting the extension SDKs to other programming languages we would also lower the barrier of entry for authoring extensions in languages such as JRuby. With  ...
```
This looks a little unclear here. Do we want to say we that we will support SDK in JRuby? If yes, then we can write something like:
```
lower the barrier of entry for authoring extensions in languages such Ruby with SDK in JRuby

dblock replied:
Fixed.

owaiskazi19 commented on _posts/2023-04-21-extensibility-roadmap.markdown:
```diff
@@ -0,0 +1,154 @@
+
+### Extensions in other technologies
+
+An extension can not be entirely implemented in a Lambda, as it must maintain a network connection with OpenSearch and shares some state information. It would be possible to create a thin client for this purpose with Lambda implementation internally to enable Lambda extensions.
```
Nit:
```suggestion
An extension can not be entirely implemented in a Lambda, as it must maintain a network connection with OpenSearch and shares some state information. It would be possible to create a lean client for this purpose with Lambda implementation internally to enable Lambda extensions.
```

owaiskazi19 commented on _posts/2023-04-21-extensibility-roadmap.markdown:
```diff
@@ -0,0 +1,154 @@
+
+### Offloading background tasks to extensions
+
+A number of background operations are currently run asynchronously in an OpenSearch cluster, including Lucene segment merges, data tier migrations (e..g hot to warm to cold), and most operations performed by ISM. These could be offloaded to dedicated or isolated compute via the extensions mechanism improving cluster scalability and availability.
```
```suggestion
A number of background operations are currently run asynchronously in an OpenSearch cluster, including Lucene segment merges, data tier migrations (e.g. hot to warm to cold), and most operations performed by ISM. These could be offloaded to dedicated or isolated compute via the extensions mechanism improving cluster scalability and availability.
```

dblock replied:
Taken!

pajuric commented:
@dblock - I have filed a ticket to put this through tech/editorial review. What is your target publishing date?

dblock commented on _posts/2023-04-21-extensibility-roadmap.markdown:
```diff
@@ -0,0 +1,154 @@
+
+### Aggregations as an extension
+
+As we work towards separating search aggregations and compute, we will refactor and extract interface and SDK components for compute separate from the search phase lifecycle. Simple search aggregations, which default to the current single pass data aggregation implementation, may still be supported for basic search use cases not requiring a sophisticated compute engine. By separating the compute  ...
```
Removed

dblock commented:
Thanks @owaiskazi19 and @seanneumann for your feedback, incorporated in https://github.com/opensearch-project/project-website/pull/1455/commits/99f859dd5639b6e4ace6e7cabbe378e559760b2c.

pajuric commented:
> 

I'm checking to see if there is an editor available since Nate is OOO the remainder of this week.

kolchfa-aws commented on _posts/2023-04-21-extensibility-roadmap.markdown:
```diff
@@ -0,0 +1,154 @@
+
+## Future
+
+Extensions will make a lot of new, big ideas possible! Below are some of our favorites.
```
```suggestion
Extensions will make a lot of new, big ideas possible! Here are some of our favorites.
```

kolchfa-aws commented on _posts/2023-04-21-extensibility-roadmap.markdown:
```diff
@@ -0,0 +1,154 @@
+
+### Search processors become extensions
+
+In [search-processor#80](https://github.com/opensearch-project/search-processor/issues/80) we proposed a new search processor pipeline. Search processors are plugins that can become extensions before plugins are deprecated.
```
```suggestion
In [search-processor #80](https://github.com/opensearch-project/search-processor/issues/80), we proposed a new search processor pipeline. Search processors are plugins that can become extensions before plugins are deprecated.
```

kolchfa-aws commented on _posts/2023-04-21-extensibility-roadmap.markdown:
```diff
@@ -0,0 +1,154 @@
+
+### Communication protocols as extensions
+
+By making communication protocols extensible, we can experiment with more performant implementations such as GRPC or no-GC implementations such as Netty, without having to modify the core engine. Check out [opensearch-sdk-java#414](https://github.com/opensearch-project/opensearch-sdk-java/issues/414#issuecomment-1471010441), which prototypes protobuf serialization.
```
```suggestion
By making communication protocols extensible, we can experiment with more performant implementations such as GRPC or no-GC implementations (for example, Netty) without having to modify the core engine. Check out [opensearch-sdk-java #414](https://github.com/opensearch-project/opensearch-sdk-java/issues/414#issuecomment-1471010441), which prototypes protobuf serialization.
```

kolchfa-aws commented on _posts/2023-04-21-extensibility-roadmap.markdown:
```diff
@@ -0,0 +1,154 @@
+
+## Help Wanted
+
+We would love your help with these projects and ideas! The best place to start is to check out [opensearch-sdk-java](https://github.com/opensearch-project/opensearch-sdk-java) and trying to implement a trivial extension on top of it.
```
```suggestion
We would love your help with these projects and ideas! The best place to start is checking out [opensearch-sdk-java](https://github.com/opensearch-project/opensearch-sdk-java) and trying to implement a trivial extension on top of it.
```

kolchfa-aws reviewed (dismissed):
LGTM. Left some suggestions.

dblock commented:
@kolchfa-aws Thank you for the edits! 
@natebower This is ready for your review.

pajuric commented on _posts/2023-03-23-extensibility-roadmap.markdown:
```diff
@@ -0,0 +1,156 @@
+date:   2023-03-23
+categories:
+  - technical-post
+meta_keywords: OpenSearch extensibility, plugins, extensions
+meta_description: The OpenSearch plugin framework presents a number of challenges for users and developers. To solve these, we've embarked on a journey to replace the plugin mechanism with a new concept of extensions.
+excerpt: The OpenSearch plugin framework presents a number of challenges for users and developers. To solve these, we've embarked on a journey to replace the plugin mechanism with a new concept of extensions.
```
Please update the meta with the following: 
meta_keywords: OpenSearch extensibility, extension framework, OpenSearch plugins, OpenSearch Dashboards
meta_description: Learn how the OpenSearch is replacing its plugin framework with a new catalog of extensions to help users and developers overcome extensibility challenges

dblock replied:
Done.

pajuric replied:
Thanks, DB

seanneumann reviewed (approved):
Thanks for the updates. LGTM!

natebower commented on _posts/2023-03-23-extensibility-roadmap.markdown:
```diff
@@ -0,0 +1,156 @@
+---
+layout: post
+title:  "OpenSearch extensibility technical roadmap"
```
How about "Technical roadmap: OpenSearch extensibility"?

natebower reviewed (changes requested):
@dblock @kolchfa-aws Please see my comments and changes and let me know if you have any questions. Once you have addressed my comments, I'll review and approve, and I'm also going to have @westondd review prior to publishing. Thanks!

dblock commented on _posts/2023-03-23-extensibility-roadmap.markdown:
```diff
@@ -0,0 +1,156 @@
+  - technical-post
+meta_keywords: OpenSearch extensibility, extensions framework, OpenSearch plugins, OpenSearch Dashboards
+meta_description: Learn how the OpenSearch is replacing its plugin framework with a new catalog of extensions to help users and developers overcome extensibility challenges.
+excerpt: The OpenSearch plugin framework presents a number of challenges for users and developers. To solve these, we've embarked on a journey to replace the plugin mechanism with a new concept of extensions.
```
I actually mean concept here. The catalog is one of the practical manifestations of it.

natebower replied:
Ok, "a new concept of extensions" is perhaps a bit awkward. I would just say "to replace the plugin mechanism with extensions", but it's up to you.

dblock commented on _posts/2023-03-23-extensibility-roadmap.markdown:
```diff
@@ -0,0 +1,156 @@
+excerpt: The OpenSearch plugin framework presents a number of challenges for users and developers. To solve these, we've embarked on a journey to replace the plugin mechanism with a new concept of extensions.
+---
+
+The core reason users choose OpenSearch is the wide range of use cases they can address with its features, such as search or log analytics. Thus, we aim to make the OpenSearch Project the preferred platform for builders by creating a vibrant and deeply integrated ecosystem of projects, features, content packs, integrations, and tools that can be found quickly, installed securely, combined to solv ...
```
- removed "cluster", internal says it all
- removed SDK words

dblock commented on _posts/2023-03-23-extensibility-roadmap.markdown:
```diff
@@ -0,0 +1,156 @@
+
+## Help Wanted
+
+We would love your help with these projects and ideas! The best place to start is checking out [opensearch-sdk-java](https://github.com/opensearch-project/opensearch-sdk-java) and trying to implement a trivial extension on top of it.
```
Expanded a bit.

We would love your help with these projects and ideas! The best way to start is by checking out [opensearch-sdk-java](https://github.com/opensearch-project/opensearch-sdk-java). Try to implement a trivial extension on top of it, or help us port an existing plugin. You could also pickup one of the issues labelled "[good first issue](https://github.com/opensearch-project/opensearch-sdk-java/issues?q=is%3Aissue+is%3Aopen+label%3A%22good+first+issue%22)" in that project. As always, please let us know how we can help by opening new issues or posting to [the forums](https://forum.opensearch.org/).

natebower replied:
"pick up" should be two words in this context, but I would use a better verb (you seem to like phrasal verbs :-)). Also, "labeled" is the American English spelling.

dblock commented on _posts/2023-03-23-extensibility-roadmap.markdown:
```diff
@@ -0,0 +1,156 @@
+
+## Introducing Extensions
+
+From the product point of view, _extensions_ are a new mechanism that provides a way to break up a monolithic, tightly coupled way of building new features in OpenSearch. Technically, extensions are a simple evolution of plugins, or plugins decoupled from their OpenSearch/Dashboards host. In practice, an extension is only different from a plugin in that it only depends on the OpenSearch/Dashboard ...
```
- took way -> method and ---
- we mean suset, aka remove, not deprecate, so left that one

natebower replied:
Ok, I would use "remove" in that case, as "sunset" is idiomatic.

dblock commented on _posts/2023-03-23-extensibility-roadmap.markdown:
```diff
@@ -0,0 +1,156 @@
+
+The proposed projects below are in order, but please note that many of these can be done in parallel.
+
+### Provide experimental SDKs and move extensibility concerns out of the monoliths
```
Replaced with cores

dblock commented on _posts/2023-03-23-extensibility-roadmap.markdown:
```diff
@@ -0,0 +1,156 @@
+
+In the plugin model, security is also a plugin. This means security can be optional, which makes it difficult for plugins to build features such as field-/document-level security (FLS/DLS), which perform data access securely. Each plugin must implement data access checks independently and correctly, which has proven itself difficult (see [security #1895](https://github.com/opensearch-project/secu ...
+
+We plan to add authentication mechanisms to the OpenSearch core (every API call or new thread/process will carry an identity), and to perform authorization when accessing data at the level of these APIs in a way that is backwards compatible with the security plugin. We will always enable authorization checks using the security plugin when called from extensions, and there will be no changes in pl ...
```
Updated.

natebower replied:
Good, but remove the first comma.

dblock commented on _posts/2023-03-23-extensibility-roadmap.markdown:
```diff
@@ -0,0 +1,156 @@
+
+OpenSearch Dashboards is a single product that includes the core platform (for example, application chrome, data and saved objects APIs), native plugins (for example, home, discover, dev tools, and stack management), and feature plugins (for example, observability, anomaly detection, maps, and alerting).
+
+The experiences in Dashboards today are not cohesive. As you move between core experiences and feature plugins, there are visual inconsistencies in the fonts, colors, layouts, and charts. Feature plugins are mostly siloed experiences and don’t render on dashboard pages. Feature plugins are built differently than the native plugins. For example, they often don’t leverage existing interfaces, such  ...
```
The current Dashboards experience is not cohesive. As you move between core experiences and feature plugins, there are visual inconsistencies in the fonts, colors, layouts, and charts. Feature plugins are mostly siloed experiences and don’t render on dashboard pages. Feature plugins are built differently than the native plugins. For example, they often don’t leverage existing interfaces, such as saved objects, to store UI-configured metadata, or render embedded components on dashboard pages. Additionally, Dashboards currently uses six libraries for rendering visualizations and offers six visualization authoring experiences with overlapping functionality, each with its own implementation.

natebower replied:
Good, but remove the comma after "metadata".

dblock commented on _posts/2023-03-23-extensibility-roadmap.markdown:
```diff
@@ -0,0 +1,156 @@
+
+### In-proc support for OpenSearch extensions
+
+We designed extensions to have a clear API boundary, yet lost the ability to run on the same JVM, adding about 10% of serialization overhead in the process. We want to give extension developers the ability to remove that overhead, at the same time providing cluster administrators with more control and improving safety and security.
```
to host extensions on the same JVM

dblock commented on _posts/2023-03-23-extensibility-roadmap.markdown:
```diff
@@ -0,0 +1,156 @@
+
+### Support dependencies between extensions
+
+In the current state, all dependencies are implicit and all versions must match, and we heavily rely on testing a distribution bundle. There’s no mechanism to know what the dependencies are, and all dependency errors are found at runtime. Any update to the dependencies requires rebuilding other plugins, even when there are no changes within the current plugin. Thus, everything is always rebuilt f ...
```
we will add the ability for extensions to depend on other extensions

dblock commented on _posts/2023-03-23-extensibility-roadmap.markdown:
```diff
@@ -0,0 +1,156 @@
+
+We are in the process of improving support for existing plugins in clients by publishing REST interfaces in the form of OpenAPI specs (see [opensearch-clients #19](https://github.com/opensearch-project/opensearch-clients/issues/19) for more information). A generator will consume the spec and output parts of a complete high-level thin client for the OpenSearch distribution.
+
+We will similarly support extensions in the clients, creating thin clients for each extension, which will be composable with the core client in every supported programming language. Extensions will publish REST interfaces in the form of OpenAPI specs, a generator will consume the spec and output a complete high-level thin client. The burden of m * n extensions and languages will be alleviated by  ...
```
- the generators don't output the same thing
- updated

dblock commented on _posts/2023-03-23-extensibility-roadmap.markdown:
```diff
@@ -0,0 +1,156 @@
+
+### Rewrite most plugins as extensions
+
+We will provide a line of sight for deprecating plugins by striving for a one-to-one correspondence of extension and plugin features with the aim to minimize the effort required to migrate to extensions. We will write a migration guide, clearly specifying the needed effort to perform the initial migration and follow-up deprecated feature replacement so that you can work it into your own task plan ...
```
We will provide a way for deprecating plugins by implementing all corresponding plugin features in extensions, with the goal of minimizing the effort required to migrate.

natebower replied:
"a way to deprecate"

dblock commented on _posts/2023-03-23-extensibility-roadmap.markdown:
```diff
@@ -0,0 +1,156 @@
+
+Assuming extensions have been widely adopted, we can deprecate the plugin APIs to be removed in the next major version of OpenSearch. We don't expect this to happen earlier than OpenSearch 4.0. Older plugin versions will continue to work with older versions of OpenSearch and receive security patches.
+
+We will replace the two distribution flavors of OpenSearch/Dashboards (currently a -min distribution without security or plugins) by a set of distribution manifests tailored for such purposes as log analytics and search. Each distribution will be represented by a manifest that can be assembled on the fly by downloading the artifacts from the public catalog for packaging purposes. We want to enabl ...
```
removed

dblock commented on _posts/2023-03-23-extensibility-roadmap.markdown:
```diff
@@ -0,0 +1,156 @@
+
+### Security isolation
+
+With extensions being designed to run in a separate VM, we can introduce multiple options for isolating extensions, such as containers and java runtimes (for example, firecracker, GraalVM, and EBPF). We can also provide a new secure default and solve the problem of JSM deprecation. We can further secure across new boundaries, ensuring all messages are encrypted in transit and all data is encrypte ...
```
With extensions being designed to run on a separate virtual machine (VM), we can introduce multiple options for isolating extensions, such as containers and Java runtimes (for example, Firecracker, GraalVM, and EBPF). We can also provide a new secure default runtime, and solve the problem of Java Security Manager (JSM) deprecation. We can further secure across new boundaries, ensuring all messages are encrypted in transit and all data is encrypted at rest.

natebower replied:
Good, but remove the comma after "runtime" (we only need to use a comma to separate two independent clauses). I would also still insert a word following "secure" in the last sentence. We can further secure what across new boundaries? Do we actually mean something like "We can extend security across new boundaries"?

dblock commented:
Thanks for a thorough review @natebower and @kolchfa-aws! Incorporated all your feedback, hopefully, this is ready for review again.

natebower reviewed (approved):
Approved with a few final comments. @westondd, Can you please review for any concerns in your area?

westondd commented on _posts/2023-03-30-technical-roadmap-opensearch-extensibility.markdown:
```diff
@@ -0,0 +1,160 @@
+
+The existing mechanism used to extend OpenSearch and OpenSearch Dashboards is a plugin framework. It provides a useful way to extend functionality, particularly when the new functionality needs access to a significant number of internal APIs. However, the plugin framework presents a number of challenges for users and developers in the areas of administration, dependency management, security, avai ...
+
+In this blog post, we'll introduce the concept of extensions and outline some proposed projects in this area.
```
There's a lot of talk below about what we "will" do. Could we add a simple disclaimer to the end of this sentence saying something like, "This blog post reflects some of our current project plans, but we want your feedback! Our plans are subject to change based on feedback, contributions, and available resources." Trying not to make it look like less of a stale legal disclaimer.

dblock replied:
I said "This blog post reflects some of our current project plans. Like all plans, they may change as we make progress, and we would love your help!"

dblock commented on _posts/2023-03-30-technical-roadmap-opensearch-extensibility.markdown:
```diff
@@ -0,0 +1,160 @@
+
+## Introducing extensions
+
+From the product point of view, extensions are a new mechanism that provides a way to break up a monolithic, tightly coupled model of building new features in OpenSearch. Technically, extensions are a simple evolution of plugins --- or plugins decoupled from their OpenSearch/Dashboards hosts. In practice, an extension is only different from a plugin in that it only depends on the OpenSearch/Dashb ...
```
We talk about major versions below, saying this won't happen before 4.0. Which is years.

dblock commented on _posts/2023-03-30-technical-roadmap-opensearch-extensibility.markdown:
```diff
@@ -0,0 +1,160 @@
+
+First, we plan to introduce an OpenSearch SDK and an OpenSearch Dashboards SDK, and refactor the OpenSearch/Dashboards cores to support them as needed. This creates both a logical and a physical separation between extensions and their hosts. You should be able to author an extension that is compatible with all minor versions of an OpenSearch/Dashboards release and to upgrade OpenSearch/Dashboards ...
+
+The SDK assumes current and future extensibility concerns from OpenSearch/Dashboards. It will contain the set of APIs that need to follow semver, significantly reducing the number of APIs that OpenSearch/Dashboards needs to worry about, because plugins only take a dependency on the SDK. With a semver-stable SDK, plugins can confidently declare that they work with, for example, `OpenSearch/Dashboa ...
```
removed will

dblock commented:
> Thanks @westondd. @dblock Can you address Weston's comments and tag me on any new/changed language? Then we should be good to go :-)

Addressed feedback.

natebower commented on _posts/2023-04-03-technical-roadmap-opensearch-extensibility.markdown:
```diff
@@ -0,0 +1,160 @@
+
+First, we plan to introduce an OpenSearch SDK and an OpenSearch Dashboards SDK, and refactor the OpenSearch/Dashboards cores to support them as needed. This creates both a logical and a physical separation between extensions and their hosts. You should be able to author an extension that is compatible with all minor versions of an OpenSearch/Dashboards release and to upgrade OpenSearch/Dashboards ...
+
+The SDK assumes current and future extensibility concerns from OpenSearch/Dashboards. It will contain the set of APIs that need to follow semver, significantly reducing the number of APIs that OpenSearch/Dashboards needs to worry about, because plugins only take a dependency on the SDK. With a semver-stable SDK, plugins can confidently declare that they work with, for example, `OpenSearch/Dashboa ...
```
```suggestion
The SDK assumes current and future extensibility concerns from OpenSearch/Dashboards. It will contain the set of APIs that need to follow semver, significantly reducing the number of APIs that OpenSearch/Dashboards needs to worry about, because plugins only take a dependency on the SDK. With a semver-stable SDK, plugins can confidently declare that they work with, for example, `OpenSearch/Dashboards >= 2.3.0` (2.3, 2.4, ... 3.0, and so on) or `~> 2.5` (any 2.x after 2.5). The SDK will provide support for integration testing against broad ranges of OpenSearch/Dashboards versions. It can begin selecting common functionality that all plugins may need, such as storing credentials or saving objects, and be strongly opinionated about what constitutes a semver-compatible extension point for OpenSearch/Dashboards. Instead of importing a transitive dependency (for example, `oui`), developers import an SDK namespace (for example, `sdk/ui`).
```

natebower commented on _posts/2023-04-03-technical-roadmap-opensearch-extensibility.markdown:
```diff
@@ -0,0 +1,160 @@
+
+In the plugin model, security is also a plugin. This means security can be optional, which makes it difficult for plugins to build features such as field- and document-level security (FLS/DLS), which perform data access securely. Each plugin must implement data access checks independently and correctly, which has proven to be difficult (see [security #1895](https://github.com/opensearch-project/s ...
+
+We plan to add authentication mechanisms to the OpenSearch core (every API call or new thread/process will carry an identity), and perform authorization when accessing data at the level of these APIs in a way that is backward compatible with the Security plugin. Authorization check will be enabled using the Security plugin for core APIs exposed in the SDK, and there will be no changes required in ...
```
```suggestion
We plan to add authentication mechanisms to the OpenSearch core (every API call or new thread/process will carry an identity) and perform authorization when accessing data at the level of these APIs in a way that is backward compatible with the Security plugin. Authorization checks will be enabled using the Security plugin for core APIs exposed in the SDK, and there will be no changes required in plugins to ensure backward compatibility.
```

natebower commented on _posts/2023-04-03-technical-roadmap-opensearch-extensibility.markdown:
```diff
@@ -0,0 +1,160 @@
+
+The primary reason users choose OpenSearch is the wide range of use cases they can address with its features, such as search or log analytics. Thus, we aim to make the OpenSearch Project the preferred platform for builders by creating a vibrant and deeply integrated ecosystem of projects, features, content packs, integrations, and tools that can be found quickly, installed securely, combined to s ...
+
+The existing mechanism used to extend OpenSearch and OpenSearch Dashboards is a plugin framework. It provides a useful way to extend functionality, particularly when the new functionality needs access to a significant number of internal APIs. However, the plugin framework presents a number of challenges for users and developers in the areas of administration, dependency management, security, avai ...
```
```suggestion
The existing mechanism used to extend OpenSearch and OpenSearch Dashboards is a plugin framework. It provides a useful way to extend functionality, particularly when the new functionality needs access to a significant number of internal APIs. However, the plugin framework presents a number of challenges for users and developers in the areas of administration, dependency management, security, availability, scalability, and developer velocity. To begin solving these, we've embarked on a journey to replace the OpenSearch plugin mechanism with a new catalog of _extensions_. We plan to ship two new SDKs for OpenSearch and OpenSearch Dashboards and then launch a catalog of extensions.
```

natebower commented on _posts/2023-04-03-technical-roadmap-opensearch-extensibility.markdown:
```diff
@@ -0,0 +1,160 @@
+
+## Introducing extensions
+
+From the product point of view, extensions are a new mechanism that provides a way to break up a monolithic, tightly coupled model of building new features in OpenSearch. Technically, extensions are a simple evolution of plugins --- or plugins decoupled from their OpenSearch/Dashboards hosts. In practice, an extension is only different from a plugin in that it only depends on the OpenSearch/Dashb ...
```
```suggestion
From the product point of view, extensions are a new mechanism that provides a way to break up a monolithic, tightly coupled model for building new features in OpenSearch. Technically, extensions are a simple evolution of plugins---or plugins decoupled from their OpenSearch/Dashboards hosts. In practice, an extension is only different from a plugin in that it only depends on the OpenSearch/Dashboards SDK and works with multiple versions of OpenSearch/Dashboards. We aim for the existence of many more extensions than there are plugins today, written by many more developers. Then we will sunset plugins.
```

natebower commented on _posts/2023-04-03-technical-roadmap-opensearch-extensibility.markdown:
```diff
@@ -0,0 +1,160 @@
+
+From the product point of view, extensions are a new mechanism that provides a way to break up a monolithic, tightly coupled model of building new features in OpenSearch. Technically, extensions are a simple evolution of plugins --- or plugins decoupled from their OpenSearch/Dashboards hosts. In practice, an extension is only different from a plugin in that it only depends on the OpenSearch/Dashb ...
+
+We want for extensions to become the preferred mechanism for providing functionality in OpenSearch/Dashboards. Having multiple competing implementations will produce extensions with high performance, improved security, and versatile features for all categories of users, administrators and developers.
```
```suggestion
We want for extensions to become the preferred mechanism for providing functionality in OpenSearch/Dashboards. Having multiple competing implementations will produce extensions with high performance, improved security, and versatile features for all categories of users, administrators, and developers.
```

natebower commented on _posts/2023-04-03-technical-roadmap-opensearch-extensibility.markdown:
```diff
@@ -0,0 +1,160 @@
+
+### Provide experimental SDKs and move extensibility concerns out of cores
+
+First, we plan to introduce an OpenSearch SDK and an OpenSearch Dashboards SDK, and refactor the OpenSearch/Dashboards cores to support them as needed. This creates both a logical and a physical separation between extensions and their hosts. You should be able to author an extension that is compatible with all minor versions of an OpenSearch/Dashboards release and to upgrade OpenSearch/Dashboards ...
```
```suggestion
First, we plan to introduce an OpenSearch SDK and an OpenSearch Dashboards SDK and refactor the OpenSearch/Dashboards cores to support them as needed. This creates both a logical and a physical separation between extensions and their hosts. You should be able to author an extension that is compatible with all minor versions of an OpenSearch/Dashboards release and to upgrade OpenSearch/Dashboards without having to upgrade an installed extension.
```

natebower commented on _posts/2023-04-03-technical-roadmap-opensearch-extensibility.markdown:
```diff
@@ -0,0 +1,160 @@
+
+We aim for extensions to ship independently, either more or less often than the official distribution. Users should be able to upgrade OpenSearch clusters much more easily because they won’t need to upgrade installed extensions. Additionally, with fewer versions of released extensions containing no new features, there will be less security patching.
+
+For each existing plugin that ships with the OpenSearch distribution, we will need to design a technical path forward, but this phase will present no change in user experience for the default OpenSearch/Dashboards distribution. First, we will design and implement interfaces that need to be exported via the OpenSearch/Dashboards SDK. This presents an opportunity to redesign extension points to be  ...
```
```suggestion
For each existing plugin that ships with the OpenSearch distribution, we will need to design a technical path forward, but this phase will present no change in user experience for the default OpenSearch/Dashboards distribution. First, we will design and implement interfaces that need to be exported via the OpenSearch/Dashboards SDK. This presents an opportunity to redesign extension points to be simpler and more coherent and an opportunity to refactor classes in OpenSearch/Dashboards. Plugins remove the dependency on OpenSearch/Dashboards and add a dependency on the SDK, reimplement the calls, and are then available as extensions. Second, to migrate the entire Security plugin into the core, we will need to add support for authorization in REST handlers, implement authorized request forwarding in the REST layer, add support for asynchronous operations and background tasks, and add system index support to allow extensions to reserve system indexes. Finally, the distribution mechanisms can begin picking up the latest available version of an extension and releasing together instead of rebuilding everything.
```

natebower replied:
I would include a noun or pronoun between "releasing" and "together" in the last sentence. Releasing what together?

natebower commented on _posts/2023-04-03-technical-roadmap-opensearch-extensibility.markdown:
```diff
@@ -0,0 +1,160 @@
+
+### OpenSearch Dashboards cohesion through interfaces
+
+OpenSearch Dashboards is a single product that includes the core platform (e.g. the application chrome, data, and saved objects APIs), native plugins (e.g. home, discover, dev tools, and stack management), and feature plugins (e.g. observability, anomaly detection, maps, and alerting).
```
```suggestion
OpenSearch Dashboards is a single product that includes the core platform (for example, the application chrome, data, and saved objects APIs), native plugins (for example, Home, Discover, Dev Tools, and Stack Management), and feature plugins (for example, Observability, Anomaly Detection, Maps, and Alerting).
```

natebower commented on _posts/2023-04-03-technical-roadmap-opensearch-extensibility.markdown:
```diff
@@ -0,0 +1,160 @@
+
+### Security isolation
+
+With extensions being designed to run on a separate virtual machine (VM), we can introduce multiple options for isolating extensions, such as containers and Java runtimes (e.g. Firecracker, GraalVM, and EBPF). We can also provide a new secure default runtime, and solve the problem of Java Security Manager (JSM) deprecation. We can further secure across new boundaries, ensuring all messages are en ...
```
```suggestion
With extensions being designed to run on a separate virtual machine (VM), we can introduce multiple options for isolating extensions, such as containers and Java runtimes (for example, Firecracker, GraalVM, and EBPF). We can also provide a new secure default runtime and solve the problem of Java Security Manager (JSM) deprecation. We can further secure across new boundaries, ensuring all messages are encrypted in transit and all data is encrypted at rest.
```

natebower replied:
Last sentence: Either "We can further extend security across new boundaries" or "We can further secure [noun] across new boundaries".

natebower commented on _posts/2023-04-03-technical-roadmap-opensearch-extensibility.markdown:
```diff
@@ -0,0 +1,160 @@
+
+### Offloading background tasks to extensions
+
+A number of background operations currently run asynchronously in an OpenSearch cluster, including Lucene segment merges, data tier migrations (e.g. hot to warm to cold), and most operations performed by Index State Management (ISM). These can be offloaded to dedicated or isolated compute instances through the extensions mechanism, improving cluster scalability and availability.
```
```suggestion
A number of background operations currently run asynchronously in an OpenSearch cluster, including Lucene segment merges, data tier migrations (for example, hot to warm to cold), and most operations performed by Index State Management (ISM). These can be offloaded to dedicated or isolated compute instances through the extensions mechanism, improving cluster scalability and availability.
```

natebower commented on _posts/2023-04-03-technical-roadmap-opensearch-extensibility.markdown:
```diff
@@ -0,0 +1,160 @@
+
+## Help wanted
+
+This blog post reflects some of our current project plans. Like all plans, they may change as we make progress, and we would love your help! The best way to start is by checking out [opensearch-sdk-java](https://github.com/opensearch-project/opensearch-sdk-java). Try to implement a trivial extension on top of it, or help us port an existing plugin. You could also pickup one of the issues labelled ...
```
```suggestion
This blog post reflects some of our current project plans. Like all plans, they may change as we make progress, and we would love your help! The best way to start is by checking out [opensearch-sdk-java](https://github.com/opensearch-project/opensearch-sdk-java). Try to implement a trivial extension on top of it, or help us port an existing plugin. You could also pick up one of the issues labeled "[good first issue](https://github.com/opensearch-project/opensearch-sdk-java/issues?q=is%3Aissue+is%3Aopen+label%3A%22good+first+issue%22)" in that project or start with one of the ideas we mentioned above. As always, please let us know how we can help by opening new issues or posting to [the forums](https://forum.opensearch.org/).
```

dblock commented:
> @dblock One more set of commits and two comments to address, and then we should be good to go.

All done.

natebower commented:
@dblock Apologies, I would normally do them as part of one larger review. Thanks, though! I think we're good to go.

natebower reviewed (dismissed):
@pajuric This is ready to publish.

dtaivpp commented:
@pajuric we may want to give the Slack one a day to ride before publishing this.

pajuric commented:
@dtaivpp - Completely agree

@dblock - Let's set the publish date to April 5 and plan to get it out before noon PST.

dblock commented:
@natebower @pajuric @dtaivpp @seanneumann thanks all! I updated the file date to 4/5.

nateynateynate reviewed (commented):
LGTM.

dtaivpp reviewed (approved):
Looks good!
